### PR TITLE
WIP: Month of year type

### LINKED ISSUE #134: `MonthOfYear` data type
We already have `DayOfWeek`; this structure (and relevant instances and functions) would be handy as well
```
data MonthOfYear
  = January
  | February
  | March
  | April
  | May
  | June
  | July
  | August
  | September
  | October
  | November
  | December
```

### PR DESCRIPTION
**Pull request by NorfairKing**
This PR exists to show that I'm happy to do the work necessary to resolve #134

Current status: The code compiles, but with warnings and failing tests.

At this point I want to acknowledge that resolving #134 is a nontrivial amount of work.
The internals of the time library become slightly more complex *in order to prevent footguns and runtime errors for users*.
There are also minor performance improvements in some places and regressions in others.

I understand that, as the maintainer @AshleyYakeley, you are in your right to close this PR without comment or even ignore it entirely.
However, I would strongly urge you to consider how many footguns and runtime issues this change can prevent, in the interest of the library's users.

I also understand that this PR will likely cause some ruffled feathers, so I want to make it extra clear that I mean no offense toward anyone who disagrees with my technical viewpoint on this matter.
I only want to make an effort towards a less dangerous technical solution.

### CODE REVIEW

File: lib/Data/Time/Calendar/MonthDay.hs [hunk 1]
```diff
@@ -74,35 +60,27 @@ dayOfYearToMonthAndDay isLeap yd =
...
-    | yd > n = (\(m, d) -> (m + 1, d)) (findMonthDay ns (yd - n))
-findMonthDay _ yd = (1, yd)
+    | yd > n = (\(m, d) -> (succ m, d)) (findMonthDay ns (yd - n))
+findMonthDay _ yd = (January, yd)
 
 -- | The length of a given month in the Gregorian or Julian calendars.
 -- First arg is leap year flag.
 monthLength :: Bool -> MonthOfYear -> DayOfMonth
-monthLength isLeap month' = monthLength' isLeap (clip 1 12 month')
+monthLength isLeap month = case month of
```
Comment (NorfairKing) [inline lib/Data/Time/Calendar/MonthDay.hs:71]: This function may be slightly more performant now.

File: test/main/Test/Arbitrary.hs [hunk 2]
```diff
@@ -23,6 +23,11 @@ instance Arbitrary FirstWeekType where
 
 deriving instance Show FirstWeekType
 
+instance Arbitrary MonthOfYear where
+    arbitrary = choose (1, 12) `suchThatMap` parseMonthOfYearIndex
+    shrink January = []
+    shrink _ = [January]
```
Comment (NorfairKing) [inline test/main/Test/Arbitrary.hs:29]: More accurate shrinking may be in order.

**Comment by AshleyYakeley** [comment]
#134 is already closed.

**Comment by AshleyYakeley** [comment]
This makes it easier to work with months-of-year as names, while more difficult to work with them as numbers. I believe the latter is the more common case, so a net negative.

**Comment by AshleyYakeley** [comment]
The "footgun" in the current design (using a month-of-year outside the range 1..12) is already obvious to anyone using the library, and is mirrored by "footguns" for day-of-month, hour-of-day, etc., which this PR does not (and of course should not) address.

**Comment by AshleyYakeley** [comment]
Finally, this PR would break a lot of existing code.

**Comment by NorfairKing** [comment]
@AshleyYakeley I'd be happy to close this and make a new PR with only the following 

* Turn all occurrences of the `MonthOfYear` type back into `Int`
* Add the `parseMonthOfYearIndex and `monthOfYearIndex` functions.
* Make a new `MonthOfYear` sum type

That way backward incompatibility is minimised because `fromGregorian`/`toGregorian` remain unchanged, and users don't need to define their own `MonthOfYear` type that then clashes in the namespace with the type synonym.

Would such a PR be more acceptable?

**Pull request closed without merging**